src/crypto/crypto_util.h

```diff
@@ -0,0 +1,680 @@
+  }
+
+ private:
+  std::vector<v8::Local<v8::Value>> list_;
+  Environment* env_;
+};
+
+template <class TypeName>
+void array_push_back(const TypeName* md,
+                     const char* from,
+                     const char* to,
+                     void* arg) {
+  static_cast<CipherPushContext*>(arg)->push_back(from);
+}
+
+inline bool IsAnyByteSource(v8::Local<v8::Value> arg) {
+  return arg->IsArrayBufferView() ||
+         arg->IsArrayBuffer() ||
+         arg->IsSharedArrayBuffer();
+}
+
+template <typename T>
+class ArrayBufferOrViewContents {
+ public:
+  ArrayBufferOrViewContents() = default;
+
+  inline explicit ArrayBufferOrViewContents(v8::Local<v8::Value> buf) {
+    CHECK(IsAnyByteSource(buf));
+    if (buf->IsArrayBufferView()) {
+      auto view = buf.As<v8::ArrayBufferView>();
+      offset_ = view->ByteOffset();
+      length_ = view->ByteLength();
+      store_ = view->Buffer()->GetBackingStore();
+    } else if (buf->IsArrayBuffer()) {
+      auto ab = buf.As<v8::ArrayBuffer>();
+      offset_ = 0;
+      length_ = ab->ByteLength();
+      store_ = ab->GetBackingStore();
+    } else {
+      auto sab = buf.As<v8::SharedArrayBuffer>();
+      offset_ = 0;
+      length_ = sab->ByteLength();
+      store_ = sab->GetBackingStore();
+    }
+  }
+
+  inline const T* data() const {
+    // Ideally, these would return nullptr if IsEmpty() or length_ is zero,
+    // but some of the openssl API react badly if given a nullptr even when
+    // length is zero, so we have to return something.
+    if (size() == 0)
+      return &buf;
+    return reinterpret_cast<T*>(store_->Data()) + offset_;
+  }
+
+  inline T* data() {
+    // Ideally, these would return nullptr if IsEmpty() or length_ is zero,
+    // but some of the openssl API react badly if given a nullptr even when
+    // length is zero, so we have to return something.
+    if (size() == 0)
+      return &buf;
+    return reinterpret_cast<T*>(store_->Data()) + offset_;
+  }
+
+  inline size_t size() const { return length_; }
+
+  // In most cases, input buffer sizes passed in to openssl need to
+  // be limited to <= INT_MAX. This utility method helps us check.
+  inline bool CheckSizeInt32() { return size() <= INT_MAX; }
+
+  inline ByteSource ToByteSource() const {
+    return ByteSource::Foreign(data(), size());
+  }
+
+  inline ByteSource ToCopy() const {
+    if (size() == 0) return ByteSource();
+    char* buf = MallocOpenSSL<char>(size());
+    CHECK_NOT_NULL(buf);
+    memcpy(buf, data(), size());
+    return ByteSource::Allocated(buf, size());
+  }
+
+  inline ByteSource ToNullTerminatedCopy() const {
+    if (size() == 0) return ByteSource();
+    char* buf = MallocOpenSSL<char>(size() + 1);
+    CHECK_NOT_NULL(buf);
+    buf[size()] = 0;
+    memcpy(buf, data(), size());
+    return ByteSource::Allocated(buf, size());
+  }
+
+  template <typename M>
+  void CopyTo(std::vector<M>* vec) const { CopyBuffer(this, vec); }
```

Comment:
This could just return `std::vector<M>`, I’d say, especially considering that it does not only fill but also resize (i.e. entirely replace)) `vec`